### PR TITLE
Refactor FlagSet

### PR DESCRIPTION
**Pull request by odersky**
- Drop FlagSet#toString. This is in preparation of making Flags an opaque type
 - Drop distinction between FlagSet and FlagConjunction
 - Make FlagSet and Flag into opaque types, with extension methods
 - Refactor `is` into three methods: `isOneOf`, `isAllOf` taking a `FlagSet` and `is` taking a single `Flag`.

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala [hunk 1]
```diff
@@ -657,50 +657,50 @@ class DottyBackendInterface(outputDirectory: AbstractFile, val superCallsMap: Ma
...
+      isFinal &&  !toDenot(sym).isClassConstructor && !(sym.is(Flags.Mutable)) &&  !(sym.enclosingClass.is(Flags.Trait))
 
     def getsJavaPrivateFlag: Boolean =
       isPrivate //|| (sym.isPrimaryConstructor && sym.owner.isTopLevelModuleClass)
 
-    def isFinal: Boolean = sym is Flags.Final
+    def isFinal: Boolean = sym.is(Flags.Final)
     def isStaticMember: Boolean = (sym ne NoSymbol) &&
-      ((sym is Flags.JavaStatic) || toDenot(sym).hasAnnotation(ctx.definitions.ScalaStaticAnnot))
+      ((sym.is(Flags.JavaStatic)) || toDenot(sym).hasAnnotation(ctx.definitions.ScalaStaticAnnot))
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala]:
```suggestion
      (sym.is(Flags.JavaStatic) || toDenot(sym).hasAnnotation(ctx.definitions.ScalaStaticAnnot))
```

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 2]
```diff
@@ -285,7 +285,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       if (cls.classInfo.selfInfo ne NoType) ValDef(ctx.newSelfSym(cls))
       else EmptyValDef
     def isOwnTypeParam(stat: Tree) =
-      (stat.symbol is TypeParam) && stat.symbol.owner == cls
+      (stat.symbol.is(TypeParam)) && stat.symbol.owner == cls
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/ast/tpd.scala]:
```suggestion
      stat.symbol.is(TypeParam) && stat.symbol.owner == cls
```

File: compiler/src/dotty/tools/dotc/core/Denotations.scala [hunk 3]
```diff
@@ -303,15 +303,15 @@ object Denotations {
...
+      info.member(name).requiredSymbol("method", name, this)(_.is(Method)).asTerm
     }
     def requiredMethodRef(name: PreName)(implicit ctx: Context): TermRef =
       requiredMethod(name).termRef
 
     def requiredMethod(pname: PreName, argTypes: List[Type])(implicit ctx: Context): TermSymbol = {
       val name = pname.toTermName
       info.member(name).requiredSymbol(i"method", name, this, argTypes) { x =>
-        (x is Method) && {
+        (x.is(Method)) && {
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Denotations.scala]:
```suggestion
        x.is(Method) && {
```

File: compiler/src/dotty/tools/dotc/core/Denotations.scala [hunk 4]
```diff
@@ -1340,7 +1340,7 @@ object Denotations {
      *  enter it.
      */
     def missingHook(owner: Symbol, name: Name)(implicit ctx: Context): Symbol =
-      if ((owner is Package) && name.isTermName)
+      if ((owner.is(Package)) && name.isTermName)
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Denotations.scala]:
```suggestion
      if (owner.is(Package) && name.isTermName)
```

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 5]
```diff
@@ -1469,7 +1481,7 @@ object SymDenotations {
      */
     private def typeParamsFromDecls(implicit ctx: Context) =
       unforcedDecls.filter(sym =>
-        (sym is TypeParam) && sym.owner == symbol).asInstanceOf[List[TypeSymbol]]
+        (sym.is(TypeParam)) && sym.owner == symbol).asInstanceOf[List[TypeSymbol]]
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala]:
```suggestion
        sym.is(TypeParam) && sym.owner == symbol).asInstanceOf[List[TypeSymbol]]
```

File: compiler/src/dotty/tools/dotc/core/Symbols.scala [hunk 6]
```diff
@@ -751,7 +751,7 @@ object Symbols {
 
     /** The source or class file from which this class was generated, null if not applicable. */
     override def associatedFile(implicit ctx: Context): AbstractFile =
-      if (assocFile != null || (this.owner is PackageClass) || this.isEffectiveRoot) assocFile
+      if (assocFile != null || (this.owner.is(PackageClass)) || this.isEffectiveRoot) assocFile
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Symbols.scala]:
```suggestion
      if (assocFile != null || this.owner.is(PackageClass) || this.isEffectiveRoot) assocFile
```

File: compiler/src/dotty/tools/dotc/core/TypeErasure.scala [hunk 7]
```diff
@@ -215,7 +215,7 @@ object TypeErasure {
     }
 
   /** Is `tp` an abstract type or polymorphic type parameter that has `Any`, `AnyVal`,
-   *  or a universal trait as upper bound and that is not Java defined? Arrays of such types are
+   *  or a universal trait as upper bound and that.is(not) Java defined? Arrays of such types are
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/TypeErasure.scala]:
```suggestion
   *  or a universal trait as upper bound and that is not Java defined? Arrays of such types are
```

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 8]
```diff
@@ -4327,7 +4327,7 @@ object Types {
           case et: ExprType => true
           case _ => false
         }
-        if ((tp.cls is Trait) || zeroParams(tp.cls.primaryConstructor.info)) tp // !!! needs to be adapted once traits have parameters
+        if ((tp.cls.is(Trait)) || zeroParams(tp.cls.primaryConstructor.info)) tp // !!! needs to be adapted once traits have parameters
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Types.scala]:
```suggestion
        if (tp.cls.is(Trait) || zeroParams(tp.cls.primaryConstructor.info)) tp // !!! needs to be adapted once traits have parameters
```

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 9]
```diff
@@ -5123,7 +5123,7 @@ object Types {
     def apply(pre: Type, name: Name)(implicit ctx: Context): Boolean =
       name.isTypeName && {
         val mbr = pre.nonPrivateMember(name)
-        (mbr.symbol is Deferred) && mbr.info.isInstanceOf[RealTypeBounds]
+        (mbr.symbol.is(Deferred)) && mbr.info.isInstanceOf[RealTypeBounds]
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Types.scala]:
```suggestion
        mbr.symbol.is(Deferred) && mbr.info.isInstanceOf[RealTypeBounds]
```

File: compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala [hunk 10]
```diff
@@ -738,7 +738,7 @@ object JavaParsers {
         } else {
           if (in.token == ENUM || definesInterface(in.token)) mods |= Flags.JavaStatic
           val decls = memberDecl(start, mods, parentToken, parentTParams)
-          (if ((mods is Flags.JavaStatic) || inInterface && !(decls exists (_.isInstanceOf[DefDef])))
+          (if ((mods.is(Flags.JavaStatic)) || inInterface && !(decls exists (_.isInstanceOf[DefDef])))
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala]:
```suggestion
          (if (mods.is(Flags.JavaStatic) || inInterface && !(decls exists (_.isInstanceOf[DefDef])))
```

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 11]
```diff
@@ -83,7 +83,7 @@ class PlainPrinter(_ctx: Context) extends Printer {
    *  to the name of the owner.
    */
   protected def hasMeaninglessName(sym: Symbol): Boolean = (
-       (sym is Param) && sym.owner.isSetter    // x$1
+       (sym.is(Param)) && sym.owner.isSetter    // x$1
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]:
```suggestion
       sym.is(Param) && sym.owner.isSetter    // x$1
```

File: compiler/src/dotty/tools/dotc/transform/TransformByNameApply.scala [hunk 12]
```diff
@@ -27,7 +27,7 @@ abstract class TransformByNameApply extends MiniPhase { thisPhase: DenotTransfor
 
   /** If denotation had an ExprType before, it now gets a function type */
   protected def exprBecomesFunction(symd: SymDenotation)(implicit ctx: Context): Boolean =
-    (symd is Param) || (symd is (ParamAccessor, butNot = Method))
+    (symd.is(Param)) || symd.is(ParamAccessor, butNot = Method)
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/transform/TransformByNameApply.scala]:
```suggestion
    symd.is(Param) || symd.is(ParamAccessor, butNot = Method)
```

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 13]
```diff
@@ -467,7 +467,7 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         }
       }
       val getterPrefix =
-        if ((meth is Synthetic) && meth.name == nme.apply) nme.CONSTRUCTOR else meth.name
+        if ((meth.is(Synthetic)) && meth.name == nme.apply) nme.CONSTRUCTOR else meth.name
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]:
```suggestion
        if (meth.is(Synthetic) && meth.name == nme.apply) nme.CONSTRUCTOR else meth.name
```

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 14]
```diff
@@ -301,12 +301,12 @@ class Typer extends Namer
                       curOwner
                   effectiveOwner.thisType.select(name, defDenot)
                 }
-              if (!(curOwner is Package) || isDefinedInCurrentUnit(defDenot))
+              if (!(curOwner.is(Package)) || isDefinedInCurrentUnit(defDenot))
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]:
```suggestion
              if (!curOwner.is(Package) || isDefinedInCurrentUnit(defDenot))
```

**Comment by odersky** [comment]
I am going to merge this tomorrow since it is mostly cosmetic. It would be a pain to review.

**Pull request merged**